cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/StorageModule.java

```diff
@@ -77,15 +78,18 @@ public TransactionRunner get() {
         throw new IllegalStateException("No storage implementation is specified in the configuration file");
       }
 
+      MetricsCollectionService metricsCollectionService = injector.getInstance(MetricsCollectionService.class);
       storageImpl = storageImpl.toLowerCase();
       if (storageImpl.equals(Constants.Dataset.DATA_STORAGE_NOSQL)) {
         return new NoSqlTransactionRunner(injector.getInstance(NoSqlStructuredTableAdmin.class),
-                                          injector.getInstance(TransactionSystemClient.class));
+                                          injector.getInstance(TransactionSystemClient.class),
+                                          metricsCollectionService);
       }
 
       if (storageImpl.equals(Constants.Dataset.DATA_STORAGE_SQL)) {
         return new RetryingSqlTransactionRunner(injector.getInstance(StructuredTableAdmin.class),
-                                                injector.getInstance(DataSourceInstantiator.class).get());
+                                                injector.getInstance(DataSourceInstantiator.class).get(),
```

Comment:
I am looking at the implementation of `DataSourceInstantiator`. Seems like that implementation should be replaced with a Guice Provider. With that, then you can annotate the `RetryingSqlTransactionRunner` constructor with `@Inject` and simply call `return injector.getInstance(RetryingSqlTransactionRunner.class);` in here.

You don't have to make this change as part of this PR. But I think that's the nicer way to use Guice module, especially for the case you have in here. You basically don't need to change this class. You just need to have MetricsCollectionService as a injected field in the transaction runner implementation directly.

Reply:
👍